code/modules/mob/living/simple_animal/friendly/trader.dm

```diff
@@ -94,154 +181,285 @@
 		var/image/item_image = image(icon = initial(thing.icon), icon_state = initial(thing.icon_state))
+		product_info = products[thing]
+		if(product_info[TRADER_PRODUCT_INFO_QUANTITY] <= 0) //out of stock
+			item_image.overlays += image(icon = 'icons/hud/radial.dmi', icon_state = "radial_center")
 		items += list("[initial(thing.name)]" = item_image)
 	var/pick = show_radial_menu(user, src, items, custom_check = CALLBACK(src, .proc/check_menu, user), require_near = TRUE, tooltips = TRUE)
 	if(!pick)
 		return
-	var/path_reference = display_names[pick]
-	try_buy(user, path_reference)
+	var/obj/item/item_to_buy = display_names[pick]
 	face_atom(user)
-
-
-/**
- * Tries to buy an item from the trader
- * Arguments:
- * * user - The mob trying to buy something
- * * item_to_buy - Item that is being bought
- */
-/mob/living/simple_animal/hostile/retaliate/trader/proc/try_buy(mob/user, obj/item/item_to_buy)
-	var/cost = products[item_to_buy]
-	to_chat(user, span_notice("It will cost you [cost] credits to buy \the [initial(item_to_buy.name)]. Are you sure you want to buy it?"))
+	product_info = products[item_to_buy]
+	if(!product_info[TRADER_PRODUCT_INFO_QUANTITY])
+		to_chat(user, span_red("The item appears to be out of stock."))
+		return
+	to_chat(user, span_notice("It will cost you [product_info[TRADER_PRODUCT_INFO_PRICE]] [currency_name] to buy \the [initial(item_to_buy.name)]. Are you sure you want to buy it?"))
 	var/list/npc_options = list(
 		"Yes" = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_yes"),
 		"No" = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_no")
 	)
-	var/npc_result = show_radial_menu(user, src, npc_options, custom_check = CALLBACK(src, .proc/check_menu, user), require_near = TRUE, tooltips = TRUE)
-	if(npc_result != "Yes")
+	var/buyer_will_buy = show_radial_menu(user, src, npc_options, custom_check = CALLBACK(src, .proc/check_menu, user), require_near = TRUE, tooltips = TRUE)
+	if(buyer_will_buy != "Yes")
 		return
 	face_atom(user)
-	var/obj/item/holochip/cash
-	cash = user.is_holding_item_of_type(/obj/item/holochip)
-	if(!cash || cash.credits < products[item_to_buy])
-		say(nocashphrase)
+	if(!spend_buyer_offhand_money(user, product_info[TRADER_PRODUCT_INFO_PRICE]))
+		say(return_trader_phrase(NO_CASH_PHRASE))
 		return
-	cash.spend(products[item_to_buy])
 	item_to_buy = new item_to_buy(get_turf(user))
 	user.put_in_hands(item_to_buy)
 	playsound(src, sell_sound, 50, TRUE)
-	say(buyphrase)
+	product_info[TRADER_PRODUCT_INFO_QUANTITY] -= 1
+	say(return_trader_phrase(BUY_PHRASE))
+
+///Calculates the value of money in the hand of the buyer and spends it if it's sufficient
+/mob/living/simple_animal/hostile/retaliate/trader/proc/spend_buyer_offhand_money(mob/user, the_cost)
+	var/value = 0
+	var/obj/item/holochip/cash
+	cash = user.is_holding_item_of_type(/obj/item/holochip)
+	if(cash)
+		value += cash.credits
+	if((value >= the_cost) && cash)
+		cash.spend(the_cost)
+		return TRUE
```

Comment:
This will return TRUE, even if the person doesn't have enough money.
``spend()`` returns whether or not the payment was successful, you can use that.